### PR TITLE
Feature - Add Curve to Eco Router

### PR DESCRIPTION
**Pull request by adamazad**
### Summary

Adds [Curve](https://curve.fi) to the Eco Router (?) for Ethereum Mainnet, Arbitrum One, and Gnosis Chain. The main change is `CurveTrade` class that has static method `bestTradeExactIn` to return the best trade from Curve pools or Curve Smart Router (on mainnet).

### To Test

run `npm run test` (requires) docker.

### CODE REVIEW

File: src/entities/trades/curve-trade/contracts.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+export const MAINNET_CONTRACTS = {
+  addressProvider: '0x0000000022d53366457f9d5e68ec105046fc4383',
+  router: '0xfA9a30350048B2BF66865ee20363067c66f67e58'
+} as const
+
+export const RPC_PROVIDER_LIST = {
+  [ChainId.MAINNET as ChainId]: 'https://mainnet.infura.io/v3/e1a3bfc40093494ca4f36b286ab36f2d',
+  [ChainId.XDAI as ChainId]: 'https://rpc.xdaichain.com/',
+  [ChainId.ARBITRUM_ONE as ChainId]: 'https://arb1.arbitrum.io/rpc'
+}
```
Comment (adamazad) [inline src/entities/trades/curve-trade/contracts.ts:21]: Should be moved to a default namespace

File: src/entities/trades/curve-trade/contracts.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+/**
+ *  Construct a new read-only Provider
+ */
+export const getProvider = (chainId: ChainId) => {
+  const host = RPC_PROVIDER_LIST[chainId]
+  return new JsonRpcProvider(host)
+}
```
Comment (adamazad) [inline src/entities/trades/curve-trade/contracts.ts:29]: Same comment as above

File: src/entities/trades/curve-trade/CurveTrade.ts [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+  }
+
+  return tokenList.findIndex(({ address }) => address.toLowerCase() == tokenAddress.toLowerCase())
+}
+
+function getCurveToken(tokenAddress: string, chainId: ChainId = ChainId.MAINNET) {
+  const tokenList = CURVE_TOKENS[chainId as keyof typeof CURVE_TOKENS]
+
+  return Object.values(tokenList).find(token => token.address.toLowerCase() === tokenAddress.toLowerCase())
+}
```
Comment (adamazad) [inline src/entities/trades/curve-trade/CurveTrade.ts:59]: Move to utils

File: src/entities/trades/curve-trade/CurveTrade.ts [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+    const hasUnderlyingTokenOut = underlyingTokens?.some(
+      token => token.address.toLowerCase() === tokenOutAddress.toLowerCase()
+    )
+
+    return (
+      (hasTokenIn || hasUnderlyingTokenIn || hasMetaTokenIn) &&
+      (hasTokenOut || hasUnderlyingTokenOut || hasMetaTokenOut)
+    )
+  })
+}
```
Comment (adamazad) [inline src/entities/trades/curve-trade/CurveTrade.ts:103]: Move to utils

File: src/entities/trades/curve-trade/CurveTrade.ts [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+export class CurveTrade extends Trade {
+  /**
+   * An address the EOA must approve to spend its tokenIn
+   */
+  public readonly approveAddress: string
+  /**
+   * The Unsigned transaction
+   */
+  public readonly transactionRequest: UnsignedTransaction
+
```
Comment (adamazad) [inline src/entities/trades/curve-trade/CurveTrade.ts:118]: Add fee model property to base `Trade` class

File: src/entities/trades/curve-trade/CurveTrade.ts [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+  public static async canRoute(tokenIn: Token, tokenOut: Token): Promise<boolean> {
+    // Validations
+    const chainId: ChainId | undefined =
+      tokenIn instanceof Token ? tokenIn.chainId : tokenOut instanceof Token ? tokenOut.chainId : undefined
+    invariant(chainId !== undefined && RoutablePlatform.CURVE.supportsChain(chainId), 'CHAIN_ID')
+    invariant(!tokenIn.equals(tokenOut), 'CURRENCY')
+    const routerContract = new Contract(MAINNET_CONTRACTS.router, CURVE_ROUTER_ABI, getProvider(ChainId.MAINNET))
+    // Get the the router an pass arguments
+    return routerContract.can_route(tokenIn.address, tokenOut.address)
+  }
```
Comment (adamazad) [inline src/entities/trades/curve-trade/CurveTrade.ts:197]: Remove

File: src/entities/trades/curve-trade/contracts.ts [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+    const [routes, indices, expectedAmountOut] = res
+
+    return {
+      expectedAmountOut,
+    indices,
+    routes
+}
+  } catch (e) {
+    // console.log(e)
```
Comment (karczuRF) [inline src/entities/trades/curve-trade/contracts.ts:128]: Shouldn't delete comments and handle error?

Comment (adamazad) [reply src/entities/trades/curve-trade/contracts.ts:128]: > Shouldn't delete comments and handle error?

Contract throws an exception if there's no routing information. Are there better ways to handle this case?

Comment (karczuRF) [reply src/entities/trades/curve-trade/contracts.ts:128]: That's ok. So in this case I would leave just short comment explaining that contract throws an exception or throw new error here. This commented if-else statement is imo confusing.

**Review by karczuRF — APPROVED** [review]

**Pull request merged**